README.md
**Launched on Goerli:** [Mint](https://person-al.github.io/out/pathfinder/) | [View Collection on OpenSea](https://testnets.opensea.io/collection/pathfinder)

# Poetry through history

Create a poem with strangers. Play with the boundaries of on-chain history and ownership.

You may own or you may create. You cannot do both.

## How does it work? (aka the Mechanics)

PoemPathfinder contains a diamond-shaped poem which, in itself, contains 21600 potential poems:

![](./assets/diamond.svg)

Each of the 7 mintable NFTs represent a choice: left or right? As tokens are transferred and held, the path before them changes. The more owners a token has, the more likely it is to stray off the strict left-right path. And the longer a token is held, the more the future hides from view. To take the next step, a token must be burned and the image for all tokens is updated to reflect the current path. Mint, hold, transfer, sell, or burn, it's up to you! Let's see what we make together.

### Rules

- there are 7 total NFTs available
- you may mint a maximum of one token
- you may hold up to 3 tokens at a time
- minting, transferring, and holding all impact the path we take down the diamond
- burning a tokens solidifies our next step

## Architecture

Here's a quick diagram of the contracts involved:
[![](https://mermaid.ink/img/pako:eNptVdtu1DAQ_ZVRHiiIrUQR4rJCQqWtoBKiqAUq0L7MxpONWce2bKfLUvrvjJ2LE7R5SuwztzNnJvdFaQQVy6JU6P25xI3DZqWBn4vrs1fPT07h7d_jY_hiqFn2RyAbq6ghHTBIozv01U7jWtEUHRyWWyiNji8BzE6T87W0gFoAliV5P7NdwpUl_ZOsJSU1-MA4dKLDXHLs58-m7ltPAoKBnQy1cLiDDhHMlrSfGs39Tm7evD6ZOnRmjypI8iB1IFdhSR345m7T4a5JC3Ix1-7iPXp6-eLwXf6ecRKz-3Z9mTjYUIiuq1aXkcg-6QjsTL4Sc8AO4kl3l7o0O4d76K7i49uUdV-ARpV9Q2UcBDaUepMNrj5_-tERGa9rZrvGkGB-AZruyIEgq8ye-jY8TPPoFXI_-DtmD1oojt9wAhxoETWgfUUufaxbp7vTyMKYdxZGUswkwafga9MqAZewaQNztgdTQajJU67s3eMnGX_qKN4DWuvM3YwATVFx6NiHyya_Ws-OwVPZOhn24KTfjh5n1Ua2x1LXrZea3YEyG1lmAnwwjuuPKaxZHGDZaJE-2b_jkQGLoV7kDsSrWkYzWXK6UlsudJC2oFIKyuBdTVzfjmBjuJzfYQEUJsGZK09bIssUTgpXckvjlPRXwk_NlOgLyUHHWEdKDeEmZQ5Cq5xpBh0sc9v7TmcDw1Jy7PZ_i2EeslI4HbNl9XE8h-cYMDe3RM1KCE42HaPDZhGsIGDdolIzMXR43lYsBeYMPROLG4o6StXNwKPSGiNkFUUCPi66-B6Fl_g5rIw86qM-xkGggIKLSPN-8_0DCyTOAo8-a36yPnO6Y1BOwNggG_knSlr6sd4jn2SWKpkXPcsqrpb7rB0l1y6KP8QJdyljHzXKTfeM9DVa8hl-zOGsZBk-SoQ3cUXEzn00Ac5q5P0c_Q8pHQjf78ZJBut9iKNhBg7YmFXBi-pA1BTrFkNZJ53dGse9ORhteIpF0ZBrUAr-m6Woq4JV0tCqWPKroApbFVbFSj8wtLXcFboQkokslhUqT4sC22Bu9roslsG1NID6n-KIsqh_GjN8P_wDZEZc4g)](https://mermaid.live/edit#pako:eNptVdtu1DAQ_ZVRHiiIrUQR4rJCQqWtoBKiqAUq0L7MxpONWce2bKfLUvrvjJ2LE7R5SuwztzNnJvdFaQQVy6JU6P25xI3DZqWBn4vrs1fPT07h7d_jY_hiqFn2RyAbq6ghHTBIozv01U7jWtEUHRyWWyiNji8BzE6T87W0gFoAliV5P7NdwpUl_ZOsJSU1-MA4dKLDXHLs58-m7ltPAoKBnQy1cLiDDhHMlrSfGs39Tm7evD6ZOnRmjypI8iB1IFdhSR345m7T4a5JC3Ix1-7iPXp6-eLwXf6ecRKz-3Z9mTjYUIiuq1aXkcg-6QjsTL4Sc8AO4kl3l7o0O4d76K7i49uUdV-ARpV9Q2UcBDaUepMNrj5_-tERGa9rZrvGkGB-AZruyIEgq8ye-jY8TPPoFXI_-DtmD1oojt9wAhxoETWgfUUufaxbp7vTyMKYdxZGUswkwafga9MqAZewaQNztgdTQajJU67s3eMnGX_qKN4DWuvM3YwATVFx6NiHyya_Ws-OwVPZOhn24KTfjh5n1Ua2x1LXrZea3YEyG1lmAnwwjuuPKaxZHGDZaJE-2b_jkQGLoV7kDsSrWkYzWXK6UlsudJC2oFIKyuBdTVzfjmBjuJzfYQEUJsGZK09bIssUTgpXckvjlPRXwk_NlOgLyUHHWEdKDeEmZQ5Cq5xpBh0sc9v7TmcDw1Jy7PZ_i2EeslI4HbNl9XE8h-cYMDe3RM1KCE42HaPDZhGsIGDdolIzMXR43lYsBeYMPROLG4o6StXNwKPSGiNkFUUCPi66-B6Fl_g5rIw86qM-xkGggIKLSPN-8_0DCyTOAo8-a36yPnO6Y1BOwNggG_knSlr6sd4jn2SWKpkXPcsqrpb7rB0l1y6KP8QJdyljHzXKTfeM9DVa8hl-zOGsZBk-SoQ3cUXEzn00Ac5q5P0c_Q8pHQjf78ZJBut9iKNhBg7YmFXBi-pA1BTrFkNZJ53dGse9ORhteIpF0ZBrUAr-m6Woq4JV0tCqWPKroApbFVbFSj8wtLXcFboQkokslhUqT4sC22Bu9roslsG1NID6n-KIsqh_GjN8P_wDZEZc4g)

### Poem.sol

The Poem contract is the most complex and is worth a more thorough explanation:

#### Stored info:

- `path`: the current path we're taking down the diamond, as expressed by node indices.
- `currStep`: our current index in the path list.
- `_nodes`: a list of uint256s representing the entire diamond.
  - the index of a node in the list represents the node's id or index.
  - you'll note that node 0 is empty and is only used to denote the end of the graph
  - each uint256 has the following information packed into it:
    1. the node's left child's index
    2. the node's right child's index
    3. the other nodes in the same row as the node's children (known in the code as "jitters" for reasons explained later.)
    4. the string value of the node
  - this means that for node 1 in the diamond above, the following information would be packed into nodes[1]:
    1. 2 (the node's left-child, shown in the diamond as "reached")
    2. 3 (the node's right-child, aka "dropped")
    3. 0, 0, 0 (there are no other nodes in the same row as 2 & 3)
    4. The string "As he" represented as a number

#### ERC721A Overrides

- **`_beforeTokenTransfers`**:
  - if a token is being burned, take a step down the diamond
  - if we're not burning, make sure the receiver has a max of 2 tokens
- **`_afterTokenTransfers`**:
  - in all cases after a token transfer, we update our `_historicalInput` value. We'll get into this later.
  - if we're not burning the token, store the blockstamp of the token transfer. We will use this later to calculate how long a token was held before burning.
    - You'll notice that due to storage restrictions, we're storing a truncated version of the blockstamp by using the deployment blockNumber as our "epoch"
    - we are storing the blockstamp in the "Aux" property in the owner address map (`_packedAddressData`) provided by ERC721A. This works because we only care about how long a token was held by the owner who chooses to burn the token. The value in Aux will and should get overridden if a previous owner gets a new token.
  - If we're burning the last token, advance our currStep to the actual end. Because the last step is always guaranteed to end in the same place, we don't represent that step with an NFT. As such, we need to advance to the end manually once we reach the end.
- **`_extraData`**:
  - extraData is 24 bits of space ERC721A makes available to us in the token ID map (`_packedOwnerships`).
  - Here we store how many owners a token has had. We will use this later.
  - The `_extraData` hook is called in ERC721A's `transferFrom` function.

#### Internal functions

- **`_newHistoricalInput`**:
  - there were many ways to choose our path down the diamond. I chose to amalgamate the history of the contract into a pseudorandom number that could be used in the `_takeNextStep` function.
  - this function is called in the `_afterTransferTokens` hook.
  - Again, the goal here is to smash some of the transfer data into a pseudorandom number that adjusts over time in a loosely unpredictable way.
  - I know that the public can always spy on this value to decide if they want to burn on a specific day or not. I believe that (1) adding block difficulty and block number to the equation and (2) having this run whenever a transfer happens shake up enough of the public's control to make this inconvenient to force. You'd need to wait for the right week with the right address and then may get foiled by someone else transferring or burning in the middle. If someone _really_ wants to choose an outcome and goes through the effort to do so, I'm gonna call it part of the art.
  - That said, if this approach doesn't work for some reason, or if there are elements I haven't covered, please let me know so I can learn.
- **`_hiddenLevel_`**:
  - This is poorly named. I think I need to switch this to "transparency" or something.
  - This function takes in the number of blocks a given token was held by their current owner.
  - It uses that to determine the likelihood we skip this row in the poem. The idea here is that the longer a token is held, the more "faint" our path becomes.
  - The general formulat is: if the token is held for less than 4 months, there's a 0% chance we skip this row. If it's held for over a year, there's a 100% change we skip this row.
  - If we're in the middle, the percent chance is determined by the function 12x-54. This is a loosely linear line from (4,0) to (12,100).
- **`_jitterLevel`**:
  - this function returns the percent change we stray from our typical path.
  - When you look at the diamond above, there is a clear left and right child to any given node. If we were on the "his hands" node, the obvious path would be to choose either "to the clouds" or "shyly".
  - However, if a token has had a certain number of owners, there's a chance we pick a random other node in that row: either "joyously" or "towards his shoes".
  - This function looks at the number of owners a token has had and returns that percentage, ranging from 0 to a max of 30%.
  - note to self: maybe I need to up that to 60%? So that makes an even 20% chance per node?
- **`_getCurrIndex`**:
  - this function returns the current index we're on by checking path.
  - this is straightforward in most cases. However, let's say someone held a token for too long and we skipped a given row. In that case, we want `_getCurrIndex` to pick a potential index we _could_ be on so we can advance to a valid next step.
  - it uses the token's owner's address as a pseudo random number to decide which index we might be on. (Again, this is gameable, but I think I'm okay with that?)
- **`_takeNextStep`**:
  - this function determines what step we should take next. It does so via the following steps:
    1. determine the likelihood we skip this step (`_hiddenLevel_`)
    2. determine the likelihood we jump off our typical path (`_jitterLevel`)
    3. determine the likelihood we choose the left child
    4. determine the likelihood we choose the right child
    5. use the historical seed to choose one of the above 4 outcomes
    6. update path and currStep to take our next step

#### Rendering

There are currently two ways the token can be rendered (it changes based on the historicalInput value). There's the diamond poem shape:

![](./assets/diamond-start.svg)

and a linear version:

![](./assets/line-start.svg)

Note that both examples are showing the initialized state, where currStep is 0 and path is [1,0,0,0,0,0,0,0,25].

(more examples in [Rendering Examples](#rendering-examples) below)

The goal was to make it easier to see the forest as well as the tree, so to speak.

## Open Questions

- Do I need a Reentrancy Guard with the way the contract is designed?
- Should I trim the ERC721A contract to just the functions I need? I can't tell if the approval functions are necessary for my use case or just a security hole.
- Are there cheaper ways to store the SVG rendering pieces?
- Are there better renderings I could do or adjustments to the rendering I should do?
- Is my randomness random enough? Are there edge cases I haven't thought of? This isn't a DeFi application, a high profile launch, or a game where fairness matters. So I don't need the randomness to be foolproof. It needs to be fun and inconvenient to game.
  - One option to simplify the mechanics is to get rid of `_historicalInput` and just use the hash of the previous block as the burn-time seed. But I like the sense of history historicalInput brings to the piece.

## Inspiration, Resources, and Kudos

PoemPathfinder draws inspiration from many sources:

- I wanted to keep as much data on-chain as possible, and create something that wouldn't mean as much if it was built off-chain. Loot and Corruption(s\*) provided helpful examples of how to achieve those goals.
- Layli Long Soldier's poem [Obligations 2](https://www.poetryfoundation.org/poems/149976/obligations-2) inspired the shape of the poem and the pathfinding concept.

The following resources were instrumental in smartcontract construction:

- Xuannü's blog posts ([1](https://cryptocoven.mirror.xyz/A622VSRm8-9oLzc8l3oFGmfnFUZQmDQ3Wx3ObhSlhsc?ref=tokendaily), [2](https://cryptocoven.mirror.xyz/0eZ0tjudMU0ByeXLlRtPzDqxGzMMZw6ldzf-HfYETW0)) on developing CryptoCoven were a crucial starting point. They combined code with context in a way that a reading a smart contract on its own couldn't do.
- The [CryptoCoven](https://www.solidlint.com/address/0x5180db8f5c931aae63c74266b211f580155ecac8), [Loot](https://www.solidlint.com/address/0xFF9C1b15B16263C61d017ee9F65C50e4AE0113D7), and [WatchfaceWorld](https://www.solidlint.com/address/0x8d3b078d9d9697a8624d4b32743b02d270334af1) smartcontracts
- w1nt3r_eth's [meta thread](https://twitter.com/w1nt3r_eth/status/1504163829881528329?s=20&t=E4WmXMQhXuYG1J7ZmXLl2Q) on developing in Solidity
- I learned a _lot_ about current best practices from [w1nt3r_eth](https://twitter.com/w1nt3r_eth), [cygaar_dev](https://twitter.com/cygaar_dev) and [nazar_ilamanov](https://twitter.com/nazar_ilamanov)'s threads breaking down smart contracts.
- The [Hardhat developer template](https://github.com/paulrberg/hardhat-template)

Thanks to CommitsVortex for the [ideas and brainstorming](https://person-al.github.io/%F0%9F%8C%B0/2022/06/08/a-meeting-on-nft-art.html). Thanks to w1nt3r_eth, nazar_ilamanov, Doggo, and [thenerdassassin](https://twitter.com/thenerdassassin) for your encouragement and Solidity expertise. And thanks to meta for being the first to mint PoemPathfinder on the testnet!

## Rendering Examples

### Diamond, path with skips and a jitter: [1,3,5,7,0,0,0,0,25] currStep: 8

Note that white denotes a chosen step and crossed out darker nodes represent a row being skipped.

![](./assets/diamond-path-skips.svg)

### Line, path with skips: [1,3,0,8,11,18,0,0,25] currStep: 6

Note that only the chosen path is displayed and hidden and future steps are both whited out.

![](./assets/line-skips.svg)

## Developer Tips

The following tips come from the [Hardhat Template repo](https://github.com/paulrberg/hardhat-template). For more details, I recommend checking out their instructions.

### Compile

Compile the smart contracts with Hardhat:

```sh
$ yarn compile
```

### TypeChain

Compile the smart contracts and generate TypeChain bindings:

```sh
$ yarn typechain
```

### Test

Run the tests with Hardhat:

```sh
$ yarn test
```

### Lint Solidity

Lint the Solidity code:

```sh
$ yarn lint:sol
```

### Lint TypeScript

Lint the TypeScript code:

```sh
$ yarn lint:ts
```

### Coverage

Generate the code coverage report:

```sh
$ yarn coverage
```

### Report Gas

See the gas usage per unit test and average gas per method call:

```sh
$ REPORT_GAS=true yarn test
```

### Clean

Delete the smart contract artifacts, the coverage reports and the Hardhat cache:

```sh
$ yarn clean
```

### Deploy

Deploy the contracts to Hardhat Network:

```sh
$ yarn deploy --greeting "Bonjour, le monde!"
```


CHANGELOG.md
## TODOs

- [x] determine token metadata and SVG rendering
- [ ] finish all unit tests
- [x] create script to generate all potential image outcomes
- [ ] ~~see where we can save on gas~~ (I may be leaning more heavily towards features now :/)
- [ ] ~~make royalties OpenSea compatible~~ I decided on no royalties for this project
- [ ] ~~support contract-level metadata for OpenSea: https://docs.opensea.io/docs/contract-level-metadata~~ I'll just do it manually.
- [x] deploy on testnet
- [x] remove tip functionality
- [ ] ~~make licensing clear~~ (no need, I think)
- [x] improve README
- [x] create website homepage
- [x] come up with a better project name
- [x] update repo name
- [ ] deploy on mainnet

## Potential modifications:

- [x] Add distortion to SVG based on jitter and opacity like so: https://tympanus.net/codrops/2019/02/19/svg-filter-effects-creating-texture-with-feturbulence/

## 2023-04-03

## Significant changes

- support BatchMetadataUpdate event to let OpenSea know when the metadata has changed

## Minor changes

- additional unit tests

## Gas usage

```
·---------------------------------------|---------------------------|-------------|-----------------------------·
|         Solc version: 0.8.15          ·  Optimizer enabled: true  ·  Runs: 800  ·  Block limit: 30000000 gas  │
········································|···························|·············|······························
|  Methods                              ·               17 gwei/gas               ·       1792.72 usd/eth       │
·················|······················|·············|·············|·············|···············|··············
|  Contract      ·  Method              ·  Min        ·  Max        ·  Avg        ·  # calls      ·  usd (avg)  │
·················|······················|·············|·············|·············|···············|··············
|  Poem          ·  burn                ·      57462  ·     101367  ·      71776  ·           70  ·       2.19  │
·················|······················|·············|·············|·············|···············|··············
|  Poem          ·  mint                ·      66124  ·     100324  ·      81025  ·           70  ·       2.47  │
·················|······················|·············|·············|·············|···············|··············
|  Poem          ·  transferFrom        ·      47275  ·      69213  ·      60219  ·          144  ·       1.84  │
·················|······················|·············|·············|·············|···············|··············
|  TestablePoem  ·  burn                ·      64113  ·     101371  ·      80320  ·           34  ·       2.45  │
·················|······················|·············|·············|·············|···············|··············
|  TestablePoem  ·  initialize          ·          -  ·          -  ·     434551  ·           44  ·      13.24  │
·················|······················|·············|·············|·············|···············|··············
|  TestablePoem  ·  mint                ·      83281  ·     100381  ·      91004  ·           62  ·       2.77  │
·················|······················|·············|·············|·············|···············|··············
|  TestablePoem  ·  packNode            ·      38824  ·      38884  ·      38861  ·           19  ·       1.18  │
·················|······················|·············|·············|·············|···············|··············
|  TestablePoem  ·  setHistoricalInput  ·      29064  ·      29076  ·      29071  ·           23  ·       0.89  │
·················|······················|·············|·············|·············|···············|··············
|  TestablePoem  ·  tipTheCreator       ·          -  ·          -  ·      22360  ·            2  ·       0.68  │
·················|······················|·············|·············|·············|···············|··············
|  TestablePoem  ·  transferFrom        ·      52143  ·      69257  ·      56507  ·           55  ·       1.72  │
·················|······················|·············|·············|·············|···············|··············
|  TestablePoem  ·  withdrawAllEth      ·      30291  ·      32791  ·      31541  ·            2  ·       0.96  │
·················|······················|·············|·············|·············|···············|··············
|  Deployments                          ·                                         ·  % of limit   ·             │
········································|·············|·············|·············|···············|··············
|  Poem                                 ·          -  ·          -  ·    4509581  ·         15 %  ·     137.44  │
········································|·············|·············|·············|···············|··············
|  TestablePoem                         ·          -  ·          -  ·    5866391  ·       19.6 %  ·     178.79  │
·---------------------------------------|-------------|-------------|-------------|---------------|-------------·
```

## 2023-03-19

### Minor changes

- change to poem nodes to be ready for mainnet
- add tip function

### Gas usage

·---------------------------------------|---------------------------|-------------|-----------------------------·
| Solc version: 0.8.15 · Optimizer enabled: true · Runs: 800 · Block limit: 30000000 gas │
········································|···························|·············|······························
| Methods · 21 gwei/gas · 1804.23 usd/eth │
·················|······················|·············|·············|·············|···············|··············
| Contract · Method · Min · Max · Avg · # calls · usd (avg) │
·················|······················|·············|·············|·············|···············|··············
| Poem · burn · 62457 · 99903 · 70580 · 70 · 2.67 │

test/poem/Poem.ts
import { mineUpTo } from "@nomicfoundation/hardhat-network-helpers";
import { loadFixture } from "@nomicfoundation/hardhat-network-helpers";
import type { SignerWithAddress } from "@nomiclabs/hardhat-ethers/dist/src/signer-with-address";
import fs from "fs";
import { ethers } from "hardhat";
import os from "os";
import path from "path";
import { DOMParser } from "xmldom";

import type { Poem } from "../../src/types/contracts/Poem";
import type { Poem__factory } from "../../src/types/factories/contracts/Poem__factory";
import type { Signers } from "../types";

const DESTINATION = path.join(os.tmpdir(), "poem-svg-render-");

describe("Poem", function () {
  before(async function () {
    this.signers = {} as Signers;

    const signers: SignerWithAddress[] = await ethers.getSigners();
    this.signers.admin = signers[0];
    this.signers.user = signers[1];
    this.signers.others = signers.slice(2, -1);

    this.loadFixture = loadFixture;
  });

  beforeEach(async function () {
    const { poem } = await this.loadFixture(deployPoemFixture);
    this.poem = poem;
  });

  simulateContractLifecycle(10, 500);
});

export async function deployPoemFixture(): Promise<{ poem: Poem }> {
  const signers: SignerWithAddress[] = await ethers.getSigners();
  const admin: SignerWithAddress = signers[0];

  const poemFactory: Poem__factory = (<Poem__factory>await ethers.getContractFactory("Poem")) as Poem__factory;
  const poem: Poem = <Poem>await poemFactory.connect(admin).deploy(0);
  await poem.deployed();

  return { poem };
}

const getRandomUserWithTokens = (userToToken: Map<SignerWithAddress, number[]>, admin: SignerWithAddress) => {
  const usersWithTokens = Array.from(userToToken.keys()).filter(value => {
    const tokens = userToToken.get(value) || [];
    return tokens.length > 0;
  });
  if (usersWithTokens.length == 0) {
    return [false, admin, 0];
  }
  const user = usersWithTokens[Math.floor(Math.random() * usersWithTokens.length)];
  const tokens = userToToken.get(user) || [];
  return [true, user, tokens[0]];
};

export function simulateContractLifecycle(
  outerLoopCount: number,
  innerLoopCount: number,
  debug: boolean = false,
  saveImages: boolean = false,
): void {
  it("simulate entire contract lifecycle", async function () {
    enum Options {
      MINT,
      BURN,
      TRANSFER,
      HOLD,
    }
    for (let k = 0; k < outerLoopCount; k++) {
      const userToToken = new Map<SignerWithAddress, number[]>();
      const numUsers = this.signers.others.length;
      const hasMinted = new Set<string>();
      // start by minting the first token
      let next = Options.MINT;
      let currToken = 0;
      let blockNum = 0;

      let contractPath: number[] = [];
      let tempFolder = "";
      if (saveImages) {
        tempFolder = fs.mkdtempSync(DESTINATION);
      }
      let numPics = 1;

      // TODO: if you update the 500 to 1000 you often end up with a generic failure
      for (let i = 0; i < innerLoopCount; i++) {
        if (currToken >= 7 && !getRandomUserWithTokens(userToToken, this.signers.admin)[0]) {
          // If we've minted and burned everything, we can step out of the for-loop
          break;
        }
        const a = getRandomUserWithTokens(userToToken, this.signers.admin);
        const success = a[0] as boolean;
        const randomUserWithToken = a[1] as SignerWithAddress;
        const tokenId = a[2] as number;
        const randomUser = this.signers.others[Math.floor(Math.random() * numUsers)];
        switch (next) {
          case Options.MINT: {
            if (currToken >= 7) {
              debug && console.log("We've minted all the tokens already");
              break;
            }
            if (hasMinted.has(randomUser.address)) {
              debug && console.log("User has minted before");
              break;
            }
            await this.poem.connect(randomUser).mint();
            const a = userToToken.get(randomUser) || [];
            a.push(currToken);
            userToToken.set(randomUser, a);
            currToken += 1;
            blockNum += 1;
            hasMinted.add(randomUser.address);
            debug && console.log(`MINTED TOKEN ${currToken - 1}`);
            break;
          }
          case Options.BURN: {
            // Pick a random token
            if (!success) {
              debug && console.log("Couldn't find a user with any tokens to burn");
              break;
            }
            await this.poem.connect(randomUserWithToken).burn(tokenId);
            blockNum += 1;
            debug && console.log(`BURNED TOKEN ${tokenId}`);
            const burnTokens = userToToken.get(randomUserWithToken) || [];
            burnTokens.shift();
            userToToken.set(randomUserWithToken, burnTokens);
            break;
          }
          case Options.HOLD: {
            const numBlocks = Math.floor(Math.random() * 7000 * 180); // Hold a random number of blocks up to 6 months out.
            await mineUpTo(blockNum + numBlocks);
            blockNum += numBlocks;
            debug && console.log(`HELD ${numBlocks} BLOCKS`);
            break;
          }
          case Options.TRANSFER: {
            if (!success) {
              debug && console.log("Couldn't find a user with any tokens to transfer");
              break;
            }
            const tokens = userToToken.get(randomUser) || [];
            if (tokens.length >= 3) {
              debug && console.log("Receiver had 3 tokens already");
              break;
            }
            // do the transfer
            await this.poem
              .connect(randomUserWithToken)
              .transferFrom(randomUserWithToken.address, randomUser.address, tokenId);
            // remove the token from one user's ledger and move it to the other
            const otherTokens = userToToken.get(randomUserWithToken) || [];
            otherTokens.shift();
            userToToken.set(randomUserWithToken, otherTokens);
            tokens.push(tokenId);
            userToToken.set(randomUser, tokens);
            debug &&
              console.log(`TRANSFER TOKEN ${tokenId} from ${randomUserWithToken.address} to ${randomUser.address}`);
            break;
          }
          default:
            break;
        }

        // If we want to save an image and we have an NFT minted, then save that image
        if (saveImages) {
          const newPath = [];
          for (let i = 0; i < 9; i++) {
            newPath.push(await this.poem.connect(this.signers.admin).path(i));
          }
          if (JSON.stringify(newPath) !== JSON.stringify(contractPath)) {
            await writeSvg(tempFolder, this.poem, this.signers.admin, numPics, newPath);
            contractPath = newPath;
            numPics += 1;
          }
        } else {
          // If we don't want to save the image, at least call tokenURI so we get a gas measurement on it.
          // await this.poem.connect(this.signers.admin).tokenURI(tokenId);
        }

        const nextPercentage = Math.floor(Math.random() * 100);
        if (nextPercentage <= 30) {
          next = Options.HOLD;
        } else if (nextPercentage <= 70) {
          next = Options.TRANSFER;
        } else if (nextPercentage <= 80) {
          next = Options.MINT;
        } else {
          next = Options.BURN;
        }
      }

      const path = [];
      for (let i = 0; i < 9; i++) {
        path.push(await this.poem.connect(this.signers.admin).path(i));
      }
      debug && console.log("POEM " + path);

      // re-deploy poem for next cycle
      if (outerLoopCount > 1) {
        const { poem } = await this.loadFixture(deployPoemFixture);
        this.poem = poem;
      }
    }
  });
}

async function writeSvg(
  tempFolder: string,
  poem: Poem,
  admin: SignerWithAddress,
  _fileName: number,
  contractPath: number[],
) {
  const fileName = path.join(tempFolder, _fileName + ".html");
  console.log("Rendering", fileName);
  const svg = await poem.connect(admin).getDefaultSvg();
  const currStep = await poem.connect(admin).currStep();
  fs.writeFileSync(
    fileName,
    `<html lang="en">
  <head>
  <meta http-equiv="Content-Type" 
        content="text/html; charset=utf-8">
  </head>
  <h1>${currStep} - ${JSON.stringify(contractPath)}</h1>${svg}</html>`,
  );

  // Throws on invalid XML
  new DOMParser().parseFromString(svg);
}


test/poem/Poem.render.ts
import { expect } from "chai";

export function shouldRender(): void {
  describe("Poem render requirements", function () {
    it("can parse tokenURI", async function () {
      await this.poem.connect(this.signers.admin).mint();
      const uri = await this.poem.connect(this.signers.admin).tokenURI(0);
      expect(uri).to.match(/^data:application\/json;base64,/);
      const base64 = uri.split(",")[1];
      const json = JSON.parse(Buffer.from(base64, "base64").toString());
      expect(json.name).to.equal("Gem #0");
      expect(json.image).to.match(/^data:image\/svg\+xml;base64,/);
      const svg = Buffer.from(json.image.split(",")[1], "base64").toString();
      expect(svg).to.contain("<svg xmlns");
    });

    it("renders exact SVG expected for path");
  });
}


test/poem/Poem.fixture.ts
import type { SignerWithAddress } from "@nomiclabs/hardhat-ethers/dist/src/signer-with-address";
import { ethers } from "hardhat";

import type { Poem } from "../../src/types/contracts/Poem";
import type { Poem__factory } from "../../src/types/factories/contracts/Poem__factory";

export async function deployPoemFixture(): Promise<{ poem: Poem }> {
  const signers: SignerWithAddress[] = await ethers.getSigners();
  const admin: SignerWithAddress = signers[0];

  const poemFactory: Poem__factory = <Poem__factory>await ethers.getContractFactory("Poem");
  const poem: Poem = <Poem>await poemFactory.connect(admin).deploy();
  await poem.deployed();

  return { poem };
}


tasks/svg_qa/svg_qa.ts
import type { SignerWithAddress } from "@nomiclabs/hardhat-ethers/dist/src/signer-with-address";
import fs from "fs";
import { ethers } from "hardhat";
import os from "os";
import path from "path";

import type { TestablePoem } from "../../src/types/contracts/TestablePoem";
import type { TestablePoem__factory } from "../../src/types/factories/contracts/TestablePoem__factory";

const DESTINATION = path.join(os.tmpdir(), "poem-svg-render-");

const chooseRandom = (arr: any[], num = 1) => {
  const res = [];
  for (let i = 0; i < num; ) {
    const random = Math.floor(Math.random() * arr.length);
    if (res.indexOf(arr[random]) !== -1) {
      continue;
    }
    res.push(arr[random]);
    i++;
  }
  return res;
};

export async function renderAllPossibleSVGs() {
  const _signers: SignerWithAddress[] = await ethers.getSigners();
  const admin = _signers[0];

  const poemFactory: TestablePoem__factory = <TestablePoem__factory>await ethers.getContractFactory("TestablePoem");
  const poem: TestablePoem = <TestablePoem>await poemFactory.connect(admin).deploy();
  await poem.deployed();

  const possibleValuesForPath = [
    [1],
    [0, 2, 3],
    [0, 4, 5, 6],
    [0, 7, 8, 9, 10],
    [0, 11, 12, 13, 14, 15],
    [0, 16, 17, 18, 19],
    [0, 20, 21, 22],
    [0, 23, 24],
    [25],
  ];
  const tempFolder = fs.mkdtempSync(DESTINATION);
  console.log(`images in folder: ${tempFolder}`);
  let results: number[][] = [[]];

  // for each level of depth
  possibleValuesForPath.forEach(values => {
    const newResults: number[][] = [];
    // grab each node
    values.forEach(val => {
      // for each path we have stored
      results.forEach(resultsVal => {
        // add our node to the end of the path
        const resultsValCopy = JSON.parse(JSON.stringify(resultsVal));
        resultsValCopy.push(val);
        newResults.push(resultsValCopy);
      });
    });
    // and put all our paths on the list
    results = newResults;
  });

  await writeBatchFiles(tempFolder, poem, admin, chooseRandom(results, 100));
}

async function writeBatchFiles(tempFolder: string, poem: TestablePoem, admin: SignerWithAddress, paths: number[][]) {
  const itemsPerFile = 100;
  const numPaths = paths.length;
  console.log(`gonna write files to cover ${numPaths} potential paths`);
  for (let i = 0; i < numPaths; i += itemsPerFile) {
    const fileName = path.join(tempFolder, i + ".html");
    let htmlHead = `<html lang="en">
        <head>
        <meta http-equiv="Content-Type" 
            content="text/html; charset=utf-8">
        </head>`;
    let k = i;
    for (; k < i + itemsPerFile; k++) {
      if (k >= numPaths) {
        break;
      }
      const p = paths[k];
      const currStep = Math.floor(Math.random() * 9);
      const showDiamond = Math.floor(Math.random() * 100);
      await poem.connect(admin).setPath(p);
      await poem.connect(admin).setCurrStep(currStep);
      await poem.connect(admin).setHistoricalInput(showDiamond);
      const tokenId = Math.floor(Math.random() * 7);
      const jitterLevel = Math.floor(Math.random() * 30);
      const hiddenLevel = Math.floor(Math.random() * 100);
      const svg = await poem.connect(admin).getTestableSvg(tokenId, jitterLevel, hiddenLevel);
      htmlHead +=
        `<h1>showDiamond: ${
          (showDiamond >> 3) % 2 == 1
        } - tokenId: ${tokenId} jitterLevel: ${jitterLevel} hiddenLevel: ${hiddenLevel} currStep: ${currStep}- ${p}</h1>` +
        `<div style="width: 500px;height: 500px;">${svg}</div><br/>`;
    }
    fs.writeFileSync(fileName, `${htmlHead}</html>`);
    console.log(`wrote batch ${i} with paths less than ${k}`);
  }
}

async function _writeOneFile(tempFolder: string, poem: TestablePoem, admin: SignerWithAddress, paths: number[][]) {
  const numPaths = paths.length;
  console.log(`gonna write files to cover ${numPaths} potential paths`);
  const fileName = path.join(tempFolder, "oneimage.html");
  let htmlHead = `<html lang="en">
        <head>
        <meta http-equiv="Content-Type" 
            content="text/html; charset=utf-8">
        </head>`;
  const p = paths[0];
  const currStep = Math.floor(Math.random() * 9);
  const showDiamond = 8;
  await poem.connect(admin).setPath(p);
  await poem.connect(admin).setCurrStep(currStep);
  await poem.connect(admin).setHistoricalInput(showDiamond);
  const tokenId = Math.floor(Math.random() * 7);
  const jitterLevel = 0;
  const hiddenLevel = 0;
  const svg = await poem.connect(admin).getTestableSvg(tokenId, jitterLevel, hiddenLevel);
  htmlHead +=
    `<h1>showDiamond: ${
      (showDiamond >> 3) % 2 == 1
    } - tokenId: ${tokenId} jitterLevel: ${jitterLevel} hiddenLevel: ${hiddenLevel} currStep: ${currStep}- ${p}</h1>` +
    `<div style="width: 500px;height: 500px;">${svg}</div><br/>`;
  fs.writeFileSync(fileName, `${htmlHead}</html>`);
}
